### PR TITLE
[ramda] feat: enable `strictFunctionTypes`

### PR DESCRIPTION
**Pull request by somebody1234**
Although I've tried to keep them to a minimum in this PR, there are some breaking changes. I'm fine with removing them (from this PR at least) though.
There are also some changes to tests (although I've also tried to keep those to a minimum).

After (well, if) this is merged I'm planning to first do a minor fix-up pass to add `$ExpectType` to all tests, maybe fix some types etc, then proceed with #59170 in two parts

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).
If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: n/a

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## ramda ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ramda@latest/))
was missing the following properties:
1. collectBy
1. modify
1. modifyPath
1. partialObject
1. on

<details>
<summary>as well as these 5 other properties...</summary>
<p>promap, sequence, splitWhenever, unwind, whereAny</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against cf36a4327d7535bd2204cb766e1f27d55fa9fb3e
</p>

File: types/ramda/tsconfig.json [hunk 1]
```diff
@@ -8,7 +8,7 @@
         "noImplicitAny": true,
         "noImplicitThis": false,
         "strictNullChecks": true,
-        "strictFunctionTypes": false,
+        "strictFunctionTypes": true,
```
[the remark this answers is not in the document]
Comment (somebody1234) [reply types/ramda/tsconfig.json:11]: just to catch more mistakes - iirc when k switched them on i caught a few mistakes in my new definitions in the other PR

File: types/ramda/index.d.ts [hunk 2]
```diff
@@ -448,14 +448,13 @@ export function contains<T>(a: T): (list: readonly T[]) => boolean;
...
  */
-// tslint:disable:max-line-length
 export function converge<
     TArgs extends any[],
     TResult,
     R1, R2, R3, R4, R5, R6, R7,
     RestFunctions extends Array<(...args: TArgs) => any>
 >(
-    converging: ((...args: readonly [R1, R2, R3, R4, R5, R6, R7, ...ReturnTypesOfFns<RestFunctions>]) => TResult),
+    converging: ((...args: [R1, R2, R3, R4, R5, R6, R7, ...ReturnTypesOfFns<RestFunctions>]) => TResult),
```
[the remark this answers is not in the document]
Comment (somebody1234) [reply types/ramda/index.d.ts:457]: not sure about this one specifically - but the other readonlys on the other overloads broke inference. I'm assuming function args aren't counted as readonly by ts... i might post a playground when i get back to pc - alternatively you can try adding them back andd running the tests

**Comment by somebody1234** [comment]
forgot to mention, I've changed the return type of transduce. it's a breaking change but i believe it was broken before, since it matched neither test output types nor the types given in the docs

File: types/ramda/index.d.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (somebody1234) [reply types/ramda/index.d.ts:457]: [playground](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAWxAGwBQEMBciwjIBGApgE4A0iReBx5AlIgN6JklQhlJaIBU1ANyIAvgCgxEBADdyAcxIZUaKgG1lVZQF0GgsQHp9iAM4ALOOgAmicmThlEGKKZKI0JMHOeIILiAGtEKCx-EmMfLHZEOGAg0xhjBkkZeUUNRHV0HT0pMFkyBSV0NXTSrN0JQ0QQU2jTZLzUgCYilQyytGyDIzMLNGtbe0dnV3dPb18SAKCQsIiomLiEpNz8hRb0zM6K1ebWkuKUQ+0KsUsptEjXUEhYBB8UgpIAHjFEd8QAFQBBAvCSAAeUA8lnCWDAAE9VFoKG8Pp8AEphdBQWEfRAIgCMVARTRxAGYcQAWHEAVhxADYcQB2MQAPgwcPeuwKME8eAwGAAdDzInJjHhVFicXiMYSMSSMeSMVSMdStIgAGQsNweLymPDU0RMAC8dK+SOMKIYaI+RDI4MmAoyTPRjm5vL+eB+f11+qxJtt6M5PK5fOtLv5boxTU9do+Psd-OdvyDiD1YrD4ftvv9Mdd8fdRKT4cjfqdX1jiUzUpzdrzacLGYTCIpZe9Dvz0arcZr1KS6K0YgYHNTBcDxYTiORaCgenOEEuURu0HgSBZCle6IHNiBILBkOhpvew6No+3GOxIYJxLJlPpjPRC7Zcg5jcrQqPuJPErPMoVytYY3VeAp2pLu7Gge5qWi41qqF6Eb3v2RbBh6B4Nn2zYDnBoYIVBSEBrBJYIvi9YYVGWHVlm+HvBWMHEaW6FkdByHYTWda2l2PaOJh6atvqgGjuOFxXIgM53POjyLraK6AsCYCgog4JQjComGiiB7CseYqnlKF62te7IpoRgrKc+qmvlKH4qt+zh4KS-5Dgpo5liBkBgYKkE0WxLaDu6mKkTpTZERxIZeeRdGUbhAW0b57kSqFrkoThpIdh8zG9rpbnBlxY5nLx07gLO9wLi8olFquElSTJW7ySOqK2vpoq4cSGlXsJN53q5j4ii+CJEiZX5quZiBElZnE2VAdkWg5YROcm3mVjFNaedRU0UX5uJRclM3unh82BeFcHZkx3ZJT57ERWlPGTnxAlzg8jRPEu8KFeJ67SZucnLkNSlPjV+L1R8Wm3gtzatSpuFdaq4waog+IDQaFUjaB402pNW1HXBc3Of9204WhaNIylOF4XtLFhcjAFDadU7XNlglXWs+XLvda6SRuskHml70it9zKNdpOOA7iINmeDTRQ2lsNjeB2NE7js0rYdUvuqGBMHdN9GDRVZPnZTl15bdO708VTNla9FXvRz1PyE16N6Zi-M9eDmLC0NotWoKONrYeDCJaxq0q9De7pRIE7k-xmu5cJJBNDrblFY9pUvfCb1VR97WSgi0q1jSpu-c1yXsFglgIGgEIZNVydvunco22DmoOzDwGjc7CPJq7PvwRL0Ut1jiOS27IWbd3LfZn37fBXFQ-e8FdZj7LPftorXvTz7J0ZWdWW3FrYcRwVfzR4zT3M+Vfts0D4odWXFKZ1zf047n+dgIXxdJ4Zp9SpSlc-ogf4iKljt13D4td8PJaqMAHjyWp3Ju-dgobTbqAiKHUZbKxHggxacDGKdn2vPRBfkl6Bw1mvUO111iRzEgzEqz0WYJ3RCXJ+KdSQX0IRba+JA84FyLrzT6alU5v16pZL+JNa62nsg3CCICF7BWARAwBcDwG5kgUtaBoisFwMHjAsRS04pzzkcdUmy8g4XQITTTedNt4PV3rHChxtE5tRofQtYjDJY31YQ-axtUJTcPBv1Ph1kBHoiEY5RusipEo2QUFMBISMY1gUZI2BO14rvE9lo7+atdF4JykJQh4diF6xjuQg+ikrHH1sebbmDjmG33vuwgk7i8CQy8arP2Tt-EiOiWouBEjAkxMxuE4mkS4mIASUE-hft1arzSWbJ4Ri7omNIQbOOO5KEfGqkU1kJTXKOLvmw6q1TEBCzqb7ICgj65NNUUo4JU9TldM0YM7xwyUmjKpnlSZutpn6z3obeOliqGYmWXIexayylOMfNs+2eyRa-zFi7RJOFPIDM6TclEgggA)

File: types/ramda/index.d.ts [hunk 3]
```diff
@@ -87,11 +87,11 @@ export function add(a: number): (b: number) => number;
  * Creates a new list iteration function from an existing one by adding two new parameters to its callback
  * function: the current index, and the entire list.
  */
-export function addIndex<T, U>(fn: (f: (item: T) => U, list: readonly T[]) => U[]): _.F.Curry<(a: (item: T, idx: number, list?: T[]) => U, b: readonly T[]) => U[]>;
+export function addIndex<T, U>(fn: (f: (item: T) => U, list: readonly T[]) => U[]): _.F.Curry<(a: (item: T, idx: number, list: T[]) => U, b: readonly T[]) => U[]>;
```
Comment (adispring) [inline types/ramda/index.d.ts:90]: In most time, people only want to add `index` when iterate something.

Comment (somebody1234) [reply types/ramda/index.d.ts:90]: is that related to this change though? or...

Comment (somebody1234) [reply types/ramda/index.d.ts:90]: or, well... is `undefined` ever passed as the `list` argument? otherwise it's safe to pass functions with fewer parameters: [playground](https://www.typescriptlang.org/play?jsx=0#code/MYewdgzgLgBAZgLhgHgCoD4AUBLKBTAWyVQBoZsATADyTAFcCAjPAJzIBttpiBtAXQCUMALzoYANxCURMHPgJlKVIaJgBvGAHoAVDAogYEEATxQAFtjABzGNs0wAvkA)

File: types/ramda/index.d.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (essenmitsosse) [reply types/ramda/index.d.ts:457]: Yeah makes sense.

File: types/ramda/tsconfig.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (essenmitsosse) [reply types/ramda/tsconfig.json:11]: Ok seems like a reasonable change then.

**Review by essenmitsosse — APPROVED** [review]
Looks fine to me, but I suggest a third pair of eyes take a look at the whole thing, since its not trivial

File: types/ramda/index.d.ts (hunk 3 above)
Comment (adispring) [reply types/ramda/index.d.ts:90]: Because curried function will only be invoked when you pass enough parameters, `curry` will calculate the fn's length by `Function.prototype.length`, which will ignore optional and rest parameters: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/Length#using_function_length

Comment (somebody1234) [reply types/ramda/index.d.ts:90]: but this isn't `F.Curry`-ied, right? since it's nested? idk... if you can give a testcase where it fails that'd be nice

(i tested with passing `R.curry(fn)` as `a` and it seems to still typecheck fine)

Comment (adispring) [reply types/ramda/index.d.ts:90]: Oh, I see, it's a callback, maybe ts not support well enough to strict parameters' count.
I still think it's not nessary to remove list's `?`.

**Comment by somebody1234** [comment]
uh oh... the return types for `addIndex` are wrong when passed `R.filter` - a new overload needs to be added
just tried making a single overload work for all functions... no luck unfortunately

File: types/ramda/index.d.ts (hunk 3 above)
Comment (somebody1234) [reply types/ramda/index.d.ts:90]: oops, didn't see this comment...
i had to remove the `?` otherwise 3-argument callbacks get `T[] | undefined` as third argument, which is the wrong type.
`_.F.Curry` seems to handle `?:` fine, but as said above i believe the callback isn't `_.F.Curry`-ied
(however it doesn't hurt to do more testing to make sure nothing breaks)

Comment (adispring) [reply types/ramda/index.d.ts:90]: OK

**Comment by somebody1234** [comment]
(note that for some reason key order changed... so i had to fix that to make tests pass)

File: types/ramda/index.d.ts [hunk 4]
```diff
@@ -2163,10 +2169,10 @@ export function toUpper(str: string): string;
  * list, successively calling the transformed iterator function and passing it an accumulator value and the
  * current value from the array, and then passing the result to the next call.
  */
-export function transduce<T, U>(xf: (arg: readonly T[]) => T[], fn: (acc: readonly U[], val: U) => U[], acc: readonly T[], list: readonly T[]): U;
-export function transduce<T, U>(xf: (arg: readonly T[]) => T[]): (fn: (acc: readonly U[], val: U) => U[], acc: readonly T[], list: readonly T[]) => U;
-export function transduce<T, U>(xf: (arg: readonly T[]) => T[], fn: (acc: readonly U[], val: U) => U[]): (acc: readonly T[], list: readonly T[]) => U;
-export function transduce<T, U>(xf: (arg: readonly T[]) => T[], fn: (acc: readonly U[], val: U) => U[], acc: readonly T[]): (list: readonly T[]) => U;
+export function transduce<T, U>(xf: (arg: readonly T[]) => T[], fn: (acc: readonly U[], val: U) => U[], acc: readonly T[], list: readonly T[]): U[];
```
Comment (adispring) [inline types/ramda/index.d.ts:2172]: According R.transduce's signature of Ramda official doc: `(c → c) → ((a, b) → a) → a → [b] → a`, I think its typescript signature is not correct.

**Review by adispring — CHANGES_REQUESTED** [review]

File: types/ramda/index.d.ts (hunk 4 above)
Comment (somebody1234) [reply types/ramda/index.d.ts:2172]: oh... `acc` and `list` should be `readonly U[]` right?
(if you mean it should be more generic than just arrays, i can add the official signature as a fallback but i suspect that'd make it *not* error on arrays when it should)

Comment (somebody1234) [reply types/ramda/index.d.ts:2172]: well... actually `U[]` can (should?) probably be replaced with `V`. not sure what else should be widened as well?

Comment (adispring) [reply types/ramda/index.d.ts:2172]: `R.transduce` can be regard as `R.compose(R.reduce, f, g, h)`, so I think `R.transduce`'s signature may be the same as `R.reduce`'s signature.

**Comment by somebody1234** [comment]
._.

**Review by adispring — CHANGES_REQUESTED** [review]

**Comment by adispring** [comment]
@types/ramda has been format just now. https://github.com/DefinitelyTyped/DefinitelyTyped/pull/59408

You can format the source code [by prettier before push](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).

File: types/ramda/index.d.ts [hunk 5]
```diff
@@ -768,7 +771,12 @@ export function endsWith<T>(subList: readonly T[]): (list: readonly T[]) => bool
  */
 export function eqBy<T, U = T>(fn: (a: T) => U, a: T, b: T): boolean;
 export function eqBy<T, U = T>(fn: (a: T) => U, a: T): (b: T) => boolean;
-export function eqBy<T, U = T>(fn: (a: T) => U): _.F.Curry<(a: T, b: T) => boolean>;
+export function eqBy<T, U = T>(
```
Comment (adispring) [inline types/ramda/index.d.ts:774]: We are not care about fn's return type, maybe we can remove U, and rewrite `eqBy`'s type like this?:

```ts
export function eqBy<T>(fn: (a: T) => void, a: T, b: T): boolean;
```

**Review by adispring — CHANGES_REQUESTED** [review]

File: types/ramda/index.d.ts (hunk 5 above)
Comment (somebody1234) [reply types/ramda/index.d.ts:774]: or maybe `=> unknown`?

Comment (adispring) [reply types/ramda/index.d.ts:774]: In this case , `unknown` equals to `void`? Both of them looks good for me.

**Comment by somebody1234** [comment]
er wait `npm test ramda` is erroring

**Comment by somebody1234** [comment]
note that removing `U` from `eqBy` will be a breaking change for anyone that specifies the type parameters of `eqBy`

**Comment by essenmitsosse** [comment]
This doesn't seem to have to with this PR, happened also on the last merge to `master` (d8cb8d8e5062d39d63e46b918cc216fa1162268f) — even though that change build correctly yesterday.

**Comment by essenmitsosse** [comment]
@somebody1234 I opened a PR that fixes the issue. #59470

**Review by adispring — APPROVED** [review]

**Comment by somebody1234** [comment]
Ready to merge
